homeassistant/components/overkiz/executor.py

```diff
@@ -0,0 +1,141 @@
+"""Class for helpers and community with the OverKiz API."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+from urllib.parse import urlparse
+
+from pyoverkiz.models import Command, Device
+
+from .coordinator import OverkizDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OverkizExecutor:
+    """Representation of an Overkiz device with execution handler."""
+
+    def __init__(self, device_url: str, coordinator: OverkizDataUpdateCoordinator):
+        """Initialize the executor."""
+        self.device_url = device_url
+        self.coordinator = coordinator
+        self.base_device_url, *_ = self.device_url.split("#")
+
+    @property
+    def device(self) -> Device:
+        """Return Overkiz device linked to this entity."""
+        return self.coordinator.data[self.device_url]
+
+    def select_command(self, *commands: str) -> str | None:
+        """Select first existing command in a list of commands."""
+        existing_commands = self.device.definition.commands
+        return next((c for c in commands if c in existing_commands), None)
+
+    def has_command(self, *commands: str) -> bool:
+        """Return True if a command exists in a list of commands."""
+        return self.select_command(*commands) is not None
+
+    def select_state(self, *states) -> str | None:
+        """Select first existing active state in a list of states."""
+        if self.device.states:
+            return next(
+                (
+                    state.value
+                    for state in self.device.states
+                    if state.name in list(states)
+                ),
+                None,
+            )
+        return None
```

Comment:
Is there a way to avoid this linear search?

Reply:
Currently not, due to how the data is returned from the API. Based on the feedback from your previous PR's, we try to limit the amount of lookups as much as possible. E.g. for sensor we only loop through it at init and after we look up values directly by key.

@tetienne @vlebourl any additional idea's?

Reply:
What about:

```Python
def select_state(self, *states) -> Optional[str]:
    """Select first existing active state in a list of states."""
    for state in states:
        if current_state := self.device.states[state]:
            return current_state.value
    return None
```

There is always a states property, so we can remove the 1st condition.

Reply:
Are you sure @tetienne? I thought that stateless devices don't have the states attribute? (or am I mistaken)

Reply:
Have a look here: https://github.com/iMicknl/python-overkiz-api/blob/main/tests/test_models.py#L64

Even if Overkiz API does not return anything, we will have an empty States structure.

Reply:
@tetienne https://github.com/iMicknl/python-overkiz-api/blob/ef6c92a8a0d7a8becd7b30438d6dd05c6119898b/pyoverkiz/models.py#L173 I looked here, which says it can be None?

Reply:
Applied your new code statement, however it is still some kind of linear search, just a bit more optimized..

Reply:
Ideally this gets optimized in the future, but its ok for now